src/Microsoft.DotNet.ImageBuilder/src/Commands/GenerateArtifactsCommand.cs

```diff
@@ -0,0 +1,173 @@
+            Func<string, TContext, string> postProcess = null)
+        {
+            foreach (TContext context in contexts)
+            {
+                string artifactPath = getArtifactPath(context);
+                if (artifactPath == null)
+                {
+                    continue;
+                }
+
+                string templatePath = getTemplatePath(context);
+                if (templatePath == null)
+                {
+                    if (Options.AllowOptionalTemplates)
+                    {
+                        continue;
+                    }
+
+                    throw new InvalidOperationException(
+                        $"The {artifactName} `{artifactPath}` does not have a {templatePropertyName} specified.");
+                }
+
+                await GenerateArtifactAsync(templatePath, artifactPath, context, getSymbols, artifactName, postProcess);
+            }
+        }
+
+        private async Task GenerateArtifactAsync<TContext>(
+            string templatePath,
+            string artifactPath,
+            TContext context,
+            Func<TContext, IReadOnlyDictionary<Value, Value>> getSymbols,
+            string artifactName,
+            Func<string, TContext, string> postProcess)
+        {
+            Logger.WriteSubheading($"Generating '{artifactPath}' from '{templatePath}'");
+
+            string generatedArtifact = await RenderTemplateAsync(templatePath, context, getSymbols);
+
+            if (generatedArtifact != null)
+            {
+                if (postProcess != null)
+                {
+                    generatedArtifact = postProcess(generatedArtifact, context);
+                }
+
+                string currentArtifact = File.Exists(artifactPath) ?
+                    await File.ReadAllTextAsync(artifactPath) : string.Empty;
+                if (currentArtifact == generatedArtifact)
+                {
+                    Logger.WriteMessage($"{artifactName} in sync with template");
+                }
+                else if (Options.Validate)
+                {
+                    int differIndex = StringExtensions.DiffersAtIndex(currentArtifact, generatedArtifact);
+                    Logger.WriteError($"{artifactName} out of sync with template{Environment.NewLine}"
+                        + $"Current:   '{GetSnippet(currentArtifact, differIndex)}'{Environment.NewLine}"
+                        + $"Generated: '{GetSnippet(generatedArtifact, differIndex)}'");
+                    _outOfSyncArtifacts.Add(artifactPath);
+                }
+                else if (!Options.IsDryRun)
+                {
+                    await File.WriteAllTextAsync(artifactPath, generatedArtifact);
+                    Logger.WriteMessage($"Updated '{artifactPath}'");
+                }
+            }
+        }
+
+        private static string GetSnippet(string source, int index) => source.Substring(index, Math.Min(100, source.Length - index));
```

Comment:
Perhaps include leading text prior to the diff in addition to the trailing text?  That will help provide more context at a glance.
```suggestion
        private static string GetSnippet(string source, int index) => source.Substring(Math.Max(0, index - 100), Math.Min(100, source.Length - index));
```

Reply:
This can get tricky because to be useful, I believe you should provide an indication where the difference begins.  With my experience so far, including the diff snippet is "good enough" to quickly know what is happening.  I did however add the diff index to the message to provide some context.  

```
                    Logger.WriteError($"{artifactName} out of sync with template starting at index '{differIndex}'{Environment.NewLine}"
                        + $"Current:   '{GetSnippet(currentArtifact, differIndex)}'{Environment.NewLine}"
                        + $"Generated: '{GetSnippet(generatedArtifact, differIndex)}'");
```

If this proves to be problematic, I suggest we refactor at that time.